### PR TITLE
Doc: Rocket.Chat 4.0 User Friendly Docker Compose Template

### PR DESCRIPTION
**Pull request by avimanyu786**
## Proposed changes

- Updated services `rocketchat` and `mongo-init-replica` command for loops to run endlessly (as per migration guide - linked in this template).
- Added migration guidelines for existing Rocket.Chat 3.x.y users in `mongo` service definition.
- Also, updated the `mongo` and `mongo-init-replica` service definitions to use MongoDB 4.2 which is now a minimum requirement to make a quick transition. I understand this has already been requested in PR https://github.com/RocketChat/Rocket.Chat/pull/23402
- Removed `--smallfiles` and `--storageEngine=mmapv1` deprecated options within the `mongod` database server command.

## Issue(s)
https://github.com/RocketChat/Rocket.Chat/issues/23341

## Further comments

Because of the recent breaking changes on the latest version, this update really helps as a migration & upgrade friendly template based on MongoDB 4.2, with necessary guidelines as comments inside the `mongo` service. Specifying the storage engine is no longer needed since `wiredTiger` is automatically set by [default since MongoDB 3.2](https://docs.mongodb.com/manual/release-notes/3.2-compatibility/#default-storage-engine-change). The `--smallfiles` configuration option was also [MMAPv1 specific](https://docs.mongodb.com/v4.0/reference/program/mongod/index.html#cmdoption-mongod-smallfiles) (deprecated).

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -5,10 +5,10 @@ services:
     image: registry.rocket.chat/rocketchat/rocket.chat:latest
     command: >
       bash -c
-        "for i in `seq 1 30`; do
+        "for (( ; ; )); do
```
[the remark this answers is not in the document]
Comment (avimanyu786) [reply docker-compose.yml:8]: Same here, and I felt a bit skeptical about this change. It was revised based on instructions in the [official migration guide](https://github.com/RocketChat/docs/blob/master/quick-start/installing-and-updating/docker-containers/mongodb-mmap-to-wiredtiger-migration.md). Perhaps to be absolutely sure about the status of deployment before it breaks out or exits?

Your guide summarizes the change as: "`for` loops slightly adjusted for both "rocketchat" and "mongo-init-replica" to run them endlessly" - as for why the change, it's not mentioned exactly. Please update the guide to clarify.

File: docker-compose.yml [hunk 2]
```diff
@@ -31,28 +31,33 @@ services:
...
-    image: mongo:4.0
+  # If you're currently running Rocket.Chat 3.x.y, use this template only after completing MMAPv1 to WiredTiger migration.
+   # a. How-to guide: https://docs.rocket.chat/quick-start/installing-and-updating/docker-containers/mongodb-mmap-to-wiredtiger-migration
+   # b. After completing migration, replace the docker-compose.yml file used in the above guide, with this one and update accordingly.
+   # c. Run "docker-compose up -d --remove-orphans" from the compose directory "/opt/rocketchat", based on step a.
+    image: mongo:4.2
     restart: unless-stopped
     volumes:
      - ./data/db:/data/db
+     # Enable (remove the #) the data dump mount below, if migration not already done, else leave it as is.
```
[the remark this answers is not in the document]
Comment (avimanyu786) [reply docker-compose.yml:42]: This statement is based on the current [`docker-compose.yml`](https://github.com/RocketChat/Rocket.Chat/blob/develop/docker-compose.yml) file in this repository whereby in default, the mount is disabled. That's why I added that line. If there are other purposes, please ensure what is necessary.

File: docker-compose.yml [hunk 3]
```diff
@@ -31,28 +31,33 @@ services:
...
   # this container's job is just run the command to initialize the replica set.
-  # it will run the command and remove himself (it will not stay running)
+  # it will run the command and terminate (it will not stay running)
   mongo-init-replica:
-    image: mongo:4.0
+    image: mongo:4.2
     command: >
       bash -c
-        "for i in `seq 1 30`; do
+        "for (( ; ; )); do
```
[the remark this answers is not in the document]
Comment (avimanyu786) [reply docker-compose.yml:54]: Replied in first comment.

File: docker-compose.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (debdutdeb) [reply docker-compose.yml:8]: The migration guide is assuming that a *working* deployment is currently in the context, thus the certainty of connection establishment.

This compose file is for **everyone**. We don't know whether the `up` command would work or not (for a multitude of reasons). If looped endlessly it'd seem like the container is running - and no way of knowing whether the container is failing or not - rendering many alerting systems useless.

File: docker-compose.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (debdutdeb) [reply docker-compose.yml:42]: That mount is for database dump. Or more specifically saying the default `mongodump` location.

See "overwrite files" section under https://docs.mongodb.com/database-tools/mongodump/#behavior

File: docker-compose.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (avimanyu786) [reply docker-compose.yml:8]: Thanks. So it looks this particular change wasn't indeed quite necessary. A bottom note on the migration guide to revert back the changes to a production ready state would have been really helpful though.

File: docker-compose.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (avimanyu786) [reply docker-compose.yml:42]: This is exactly what I've been using by default. Always had it enabled from day one. So this line can be removed in that case and the mount can just be enabled by default on this compose file.

**Pull request closed without merging**